src/GraphQL/Validation/Rules.Custom/ComplexityValidationRule.cs

```diff
@@ -1,84 +1,205 @@
+using System.Diagnostics;
+using GraphQL.Types;
 using GraphQL.Validation.Complexity;
 using GraphQL.Validation.Errors.Custom;
+using GraphQLParser;
 using GraphQLParser.AST;
 
 namespace GraphQL.Validation.Rules.Custom;
 
 /// <summary>
 /// Analyzes a document to determine if its complexity exceeds a threshold.
 /// </summary>
-public class ComplexityValidationRule : IValidationRule
+public class ComplexityValidationRule : IValidationRule, INodeVisitor
 {
     private readonly ComplexityConfiguration _complexityConfiguration;
-#pragma warning disable CS0618 // Type or member is obsolete
-    private readonly IComplexityAnalyzer _complexityAnalyzer;
-#pragma warning restore CS0618 // Type or member is obsolete
 
     /// <summary>
     /// Initializes an instance with the specified complexity configuration.
     /// </summary>
     public ComplexityValidationRule(ComplexityConfiguration complexityConfiguration)
-#pragma warning disable CS0618 // Type or member is obsolete
-        : this(complexityConfiguration, new ComplexityAnalyzer())
-#pragma warning restore CS0618 // Type or member is obsolete
     {
+        _complexityConfiguration = complexityConfiguration;
+    }
+
+    /// <inheritdoc/>
+    public ValueTask<INodeVisitor?> ValidateAsync(ValidationContext context) => new(this);
+
+    ValueTask INodeVisitor.EnterAsync(ASTNode node, ValidationContext context) => default;
+
+    ValueTask INodeVisitor.LeaveAsync(ASTNode node, ValidationContext context)
+    {
+        // Complexity analysis should run at the very end.
+        if (node is GraphQLDocument)
+        {
+            // Fast return here to avoid all possible problems with complexity analysis.
+            // For example, document may contain fragment cycles, see https://github.com/graphql-dotnet/graphql-dotnet/issues/3527
+            // Note, that ComplexityValidationRule/ComplexityAnalyzer still checks for fragment cycles since there may be no standard rules configured.
+            if (context.HasErrors)
+                return default;
+
+            try
+            {
+                using (context.Metrics.Subject("document", "Analyzing complexity"))
+                    Validate(context.Document, context.Schema);
+            }
+            catch (ComplexityError ex)
+            {
+                context.ReportError(ex);
+            }
+        }
+
+        return default;
+    }
+
+    /// <inheritdoc/>
+    public void Validate(GraphQLDocument document, ISchema schema)
```

Comment:
The previous signature was:

```cs
public void Validate(GraphQLDocument document, ComplexityConfiguration complexityParameters, ISchema? schema = null)
```

But the schema is always passed from the validation rule, and the complexity parameters are pulled from the constructor.  So this signature was slightly updated to make more sense.

Reply:
Unfortunately this name "conflicts" with `ValidateAsync` from `IValidationRule.ValidateAsync`, being that they have completely different signatures.  I just left it as-is.